include/rdma/fi_ext.h

```diff
@@ -124,6 +125,36 @@ struct fid_mem_monitor {
 };
 
 
+/* Peer provider CQ support. */
+struct fid_peer_cq;
+
+struct fi_ops_cq_owner {
+	size_t	size;
+	ssize_t (*write)(struct fid_peer_cq *cq, void *context, uint64_t flags,
+			size_t len, void *buf, uint64_t data, uint64_t tag,
+			fi_addr_t src);
```

Comment:
Which address will `src` be?

Note that there are two AVs (one is owner's, the other is peer's), so the same remote endpoint will have two addresses, one for each AV.

This function is called by peer, which write directly to the application facing CQ. so the address `src` should be the address in owner's AV. On the other hand, peer does not know that address. it only knows the address in peer AV.

Reply:
As documented, this is the address relative to the peer.  The owner is responsible for translating it if that is needed.  The details are implementation specific.

But, conceptually, there will be 3 "AVs" in use here.  The top level 'AV' translates the user's fi_addr_t value to the local/remote provider and corresponding fi_addr_t' value.  The local/remote providers each have their own AV, which may store different address formats.  The top level 'AV' will need a reverse lookup mechanism.

Reply:
My understanding from talking to Wei is that this is largely a performance improvement plan.  Is 3 AVs really an acceptable solution?  If we can get it to single pointer chase, maybe that's not horrible (although it is another cache miss in the average case).  But if there's a search or even a table indirection, we're really starting to increase latency for software abstractions.

Reply:
This is targeted as an improvement over the application accessing the 2 (or more) providers independently.  The 3 AV's mentioned above is conceptual, the implementation doesn't need to match that.  But my guess is that this is how the link provider is being implemented.  If an app uses 2 providers independently, then there's conceptually a 3rd 'AV' at the app layer.

One issue is the address formats aren't guaranteed to be the same, and one of the AVs could be implemented in HW.  With rxm, and I think with shm, there's no lookup.  The src_addr is carried in the message or retrieved directly from the connection the message arrived on.  Pairing rxm with shm would require rxm to have some reverse lookup table to convert the shm fi_addr value to some index in the rxm AV.  That table only needs to be sized relative to the number of local peers, so maybe 8-96 entries today.

To take this further, we would introduce some sort of user context that could be stored with each address. 
 It would probably work okay for SW based AV's, but HW AV's might struggle.  Such a feature would also impact shared AVs and AV_MAP options.

Maybe a peer AV can be defined with this feature, which would essentially move the reverse lookup into the shm AV.

Reply:
> Maybe a peer AV can be defined with this feature, which would essentially move the reverse lookup into the shm AV.

It would be great if one can specify the target fi_address when calling `fi_av_insert` on shm AV. In that case, the owner would generate the fi_address for an remote endpoint, then call `fi_av_insert` on shm av with that address. Then everyone will be using same address for same remote endpoint.

Reply:
That potentially forces shm (or whatever provider) to create a huge AV that's almost entirely empty.  It can also only work for if all providers are using AV_TABLE.  What's desired here is the ability to store a context with each address and return that instead.  That context could be another address or pointer.

I think there's an issue in that in a general sense, there's not necessarily a single AV associated with a CQ.  Apps are expected to figure out what AV the src_addr might be referring to based on the completion context.  But the owner doesn't have that data here.